MetaMask/metamask-extension
# fix: skip setting globally selected network when entrypoint is `wallet_switchEthereumChain` or `wallet_addEthereumChain` and there are existing pending approvals

## **Description**

Currently there is a bug in `develop` that causes pending approvals to be cleared when a dapp requests to change the dapp selected chain to a different one that is already permitted for the dapp. This was not an issue pre Chain Permissions because chain switches always required user approval. Post Chain Permissions, only chains that have not previously been permitted require user approval. This makes it possible for the dapp to change the globally selected network while there are other pending (not queued) approvals, causing those pending approvals to get cleared

This PR fixes this issue by skipping the step where the globally selected network is also changed when updating the dapp selected network IF there are pending approvals.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/28026?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Goto [test dapp](https://metamask.github.io/test-dapp/)
2. Connect and permit several chains
3. Note which chain the test dapp is on
4. Send a transaction via test dapp, but do not interact with the dialog
5. Go back to the test dapp window
6. 
``` 
await window.ethereum.request({
 "method": "wallet_switchEthereumChain",
 "params": [
  {
    chainId: "0x1" // different chain from the currently selected chain for the dapp that you permitted already
  }
],
});
```
7. See that the chainId changes to match on the test dapp page
8. See that the previous transaction approval dialog still exists (i.e. has not been cleared)

## **Screenshots/Recordings**



### **Before**



### **After**

https://github.com/user-attachments/assets/e8ac85ac-6c7a-4c01-92fc-e7de4d3f878e



## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
